airflow/models/backfill.py

```diff
@@ -59,4 +94,126 @@ class BackfillDagRun(Base):
+            name="bdr_backfill_fkey",
+            ondelete="cascade",
+        ),
+        ForeignKeyConstraint(
+            [dag_run_id],
+            ["dag_run.id"],
+            name="bdr_dag_run_fkey",
+            ondelete="set null",
+        ),
+    )
+
+
+@provide_session
+def _create_backfill(
+    *,
+    dag_id: str,
+    from_date: DateTime,
+    to_date: DateTime,
+    max_active_runs: int,
+    reverse: bool,
+    dag_run_conf: dict | None,
+    session: Session = NEW_SESSION,
+) -> Backfill | None:
+    serdag = session.get(SerializedDagModel, dag_id)
+    if not serdag:
+        raise NotFound(f"Could not find dag {dag_id}")
+
+    num_active = session.scalar(
+        select(func.count()).where(Backfill.dag_id == dag_id, Backfill.completed_at.is_(None))
+    )
+    if num_active > 0:
+        raise AlreadyRunningBackfill(
+            f"Another backfill is running for dag {dag_id}. "
+            f"There can be only one running backfill per dag."
+        )
+
+    br = Backfill(
+        dag_id=dag_id,
+        from_date=from_date,
+        to_date=to_date,
+        max_active_runs=max_active_runs,
+        dag_run_conf=dag_run_conf,
+    )
+    session.add(br)
+    session.commit()
+
+    dag = serdag.dag
+    depends_on_past = any(x.depends_on_past for x in dag.tasks)
+    if depends_on_past:
+        if reverse is True:
+            raise ValueError(
+                "Backfill cannot be run in reverse when the dag has tasks where depends_on_past=True"
+            )
+
+    backfill_sort_ordinal = 0
+    dagrun_info_list = dag.iter_dagrun_infos_between(from_date, to_date)
+    if reverse:
+        dagrun_info_list = reversed([x for x in dag.iter_dagrun_infos_between(from_date, to_date)])
+    for info in dagrun_info_list:
+        backfill_sort_ordinal += 1
+        log.info("creating backfill dag run %s dag_id=%s backfill_id=%s, info=", dag.dag_id, br.id, info)
+        dr = None
+        try:
+            dr = dag.create_dagrun(
+                triggered_by=DagRunTriggeredByType.BACKFILL,
+                execution_date=info.logical_date,
+                data_interval=info.data_interval,
+                start_date=timezone.utcnow(),
+                state=DagRunState.QUEUED,
+                external_trigger=False,
+                conf=br.dag_run_conf,
+                run_type=DagRunType.BACKFILL_JOB,
+                creating_job_id=None,
+                session=session,
+            )
+        except Exception:
+            dag.log.exception("something failed")
+            session.rollback()
+        session.add(
+            BackfillDagRun(
+                backfill_id=br.id,
+                dag_run_id=dr.id if dr else None,  # this means we failed to create the dag run
+                sort_ordinal=backfill_sort_ordinal,
+            )
+        )
+        session.commit()
+    return br
+
+
+@provide_session
+def _cancel_backfill(backfill_id, session: Session = NEW_SESSION) -> Backfill:
```

Comment:
Using `@provide_session` in internal functions generally only create more confusion without benefits. In this case I think we never want the outside to pass in a session, so we should probably just use `with create_session` inside this function to keep the `session` object bounded inside this function.

Reply:
Or alternatively this function should not call `commit` and leave the session management to the outer scope.

Reply:
i updated so i will resolve and merge. if you want to review again later and have any requested changes i'm happy to make em.